### PR TITLE
fixing filters adding sort by price desc and asc

### PR DESCRIPTION
**Pull request by Ivaylo15**

### CODE REVIEW

File: api/controllers/product.js [hunk 1]
```diff
@@ -41,17 +41,18 @@ module.exports = {
         const price = req.query.price;
 
         let filterObject = {};
```
Comment (Svetloslav15) [inline api/controllers/product.js]: Can you use const here instead of let?

File: front-end/src/services/productServices.js [hunk 2]
```diff
@@ -3,14 +3,16 @@ import { messages } from '../constants/success-messages';
...
+    getProducts: (dispatch, currPage, categoryOption, sizeOption, colorOption, priceOption) => {
 
         const categoryString = new URLSearchParams(categoryOption);
         const categoryUrl = `&${categoryString.toString()}`;
         let sizeUrl = `&size=${sizeOption.toString()}`;
         let colorUrl = `&color=${colorOption.toString()}`;
+        const priceString = new URLSearchParams(priceOption);
+        let priceUrl = `&${priceString.toString()}`;
 
-        axios.get(`${process.env.REACT_APP_BASE_URL}products?page=${currPage}&limit=12${categoryUrl}${sizeUrl}${colorUrl}`)
```
Comment (Svetloslav15) [inline front-end/src/services/productServices.js:13]: Extract `12` in a variable, in case for easier change in the future.

File: front-end/src/components/gallery/Gallery.js [hunk 3]
```diff
@@ -77,24 +95,71 @@ const PagButton = styled.button`
...
+        if (sizeOption.length > 0) {
+            sizeUrl = utilFunc.stringifyUrl('size', sizeOption);
+        }
+        if (colorOption.length > 0) {
+            colorUrl = utilFunc.stringifyUrl('color', colorOption);
+        }
+        if (priceOption) {
+            priceUrl = utilFunc.stringifyUrl('price', priceOption);
+        }
+        console.log(colorUrl)
```
Comment (Svetloslav15) [inline front-end/src/components/gallery/Gallery.js]: remove console.log

File: front-end/src/components/gallery/Gallery.js [hunk 4]
```diff
@@ -77,24 +95,71 @@ const PagButton = styled.button`
...
+
+        history.push({
```
Comment (Svetloslav15) [inline front-end/src/components/gallery/Gallery.js:149]: Should this call be in this useEffect? Isn't it better to create another useEffect with [] dependancies and it will be fired only the first time the page is rendered? What do you think?

File: front-end/src/services/utils.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+export const utilFunc = {
+    stringifyUrl: (type, value) => {
+
```
Comment (Svetloslav15) [inline front-end/src/services/utils.js]: Can the whole function be simplified to this?
```suggestion
 return `&${type}=${value}
```
Done this way it won't need modifications if you want to add new types in the future 😃

File: front-end/src/components/gallery/Gallery.js [hunk 6]
```diff
@@ -154,7 +224,7 @@ const Gallery = () => {
                             <PagButton onClick={(() => setCurrPage(index + 1))} currPage={currPage} index={index + 1} key={index}>{index + 1}</PagButton>
                         ))}
                     </div>
-                    {products.results.length >= 12 && (<Button onClick={(() => setCurrPage(currPage + 1))}>NEXT</Button>)}
```
Comment (Svetloslav15) [inline front-end/src/components/gallery/Gallery.js:157]: Extract `12` into constant variable in a file with constant like `itemsPerPage` so you easily change it on one place in the future if needed.

File: front-end/src/components/gallery/Gallery.js (hunk 4 above)
Comment (Ivaylo15) [reply front-end/src/components/gallery/Gallery.js:149]: This useEffect is for building the URL when some prop is added. I have another that is fired only on the first render to set the state if there is anything to set in the URL. I am not sure if I need to separate it because it is used to build the URL and make the call to the backend with the same data.

File: api/controllers/product.js [hunk 7]
```diff
@@ -13,7 +13,6 @@ module.exports = {
         }
     },
     editProduct: async (req, res, next) => {
-        // const id = req.params.id;
         const { id, title, category, size, color, price, image } = req.body;
         try {
```
Comment (Svetloslav15) [inline api/controllers/product.js:18]: add some validations

File: api/controllers/product.js [hunk 8]
```diff
@@ -23,10 +22,10 @@ module.exports = {
         }
     },
     deleteProduct: async (req, res, next) => {
-        const { id } = req.body;
```
Comment (Svetloslav15) [inline api/controllers/product.js:26]: What will happen if the id is invalid? You can sdd test for that.

File: front-end/src/components/editProduct/EditProduct.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+        :hover{
+            background-color: #BBB;
+            color: white;
+        }
+    }
+`;
+
+const AddProduct = (props) => {
+    const redirect = useHistory();
```
Comment (Svetloslav15) [inline front-end/src/components/editProduct/EditProduct.js]: Don't think that redirect is the right word here. Better change it to history to be more precise.
```suggestion
    const history= useHistory();
```

File: front-end/src/components/editProduct/EditProduct.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        productService.editProduct(productId, title, category, size, color, price, imageUrl)
+        redirect.push('/')
+    };
+
+
+    return (
+        <Container>
+            <h2>Add Product</h2>
+            <form>
+                <input type="text" value={title} onChange={(e) => setTitle(e.target.value)} placeholder="title" />
```
Comment (Svetloslav15) [inline front-end/src/components/editProduct/EditProduct.js:67]: You don't need to create arrow function. You can simplified it to this. And the `setTitle` function will receive the `event` and then extract `target.value` from there. Same for the other inputs.
```suggestion
                <input type="text" value={title} onChange={setTitle} placeholder="title" />
```

File: front-end/src/components/gallery/Gallery.js [hunk 11]
```diff
@@ -3,16 +3,24 @@ import styled from 'styled-components';
...
 import { useDispatch, useSelector } from "react-redux";
 import { selectFilters, selectProduct } from "../../redux/productSlice";
+import { utilFunc } from "../../services/utils";
+import { constants } from "../../constants/constants";
 
 const Container = styled.div`
     width: 80%;
     min-height: 100vh;
     background-color: whitesmoke;
     margin: 0 auto 0 auto;
```
Comment (Svetloslav15) [inline front-end/src/components/gallery/Gallery.js]: This can be simplified:
```suggestion
    margin: 0 auto;
```
When you have only two arguments first is for top and bottom and second is for right and left.

File: front-end/src/components/gallery/Gallery.js [hunk 12]
```diff
@@ -77,24 +103,69 @@ const PagButton = styled.button`
...
+        }
+        if (searchParams.get('size')) {
+            setSizeOption((sizeOption) => [...sizeOption, ...searchParams.get('size').split(',')]);
+        }
+        if (searchParams.get('color')) {
+            setColorOption((colorOption) => [...colorOption, ...searchParams.get('color').split(',')]);
+        }
+        if (searchParams.get('prise')) {
+            setPriceOption(searchParams.get('prise'))
+
```
Comment (Svetloslav15) [inline front-end/src/components/gallery/Gallery.js]: Remove this empty line. 🙈

File: front-end/src/components/gallery/Gallery.js [hunk 13]
```diff
@@ -108,13 +179,11 @@ const Gallery = () => {
...
         if (filterType === 'category') {
             if (value) {
-                filter[filterType] = value;
-                setCategoryOption(filter)
+                setCategoryOption(value)
             } else {
-                setCategoryOption({})
+                setCategoryOption('')
             }
         } else if (filterType === 'size') {
```
Comment (Svetloslav15) [inline front-end/src/components/gallery/Gallery.js]: Why don't you extract the names of the filters in a constant variable so you can reuse it here and both in the useEffect. This way you will avoid lexical mistakes.

File: front-end/src/components/productView/ProductView.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
+import { useEffect, useState } from 'react';
+import { useHistory } from 'react-router-dom';
+import styled from 'styled-components';
+import { productService } from '../../services/productServices';
+
+const Container = styled.div`
+    display: flex;
+    justify-content: space-evenly;
+    padding: 2rem 0 2rem 0;
```
Comment (Svetloslav15) [inline front-end/src/components/productView/ProductView.js:9]: This can be simplified.
```suggestion
    padding: 2rem 0;
```

File: front-end/src/components/productView/ProductView.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    width: 80%;
+    min-height: 100vh;
+    background-color: whitesmoke;
+    margin: 0 auto 0 auto;
```
Comment (Svetloslav15) [inline front-end/src/components/productView/ProductView.js]:
```suggestion
    margin: 0 auto;
```

File: front-end/src/components/productView/ProductView.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    .letf__container {
+        p {
+            margin-bottom: 1rem;
+            color: #aaa;
+            font-size: 20px;
+            font-weight: 500;
+
```
Comment (Svetloslav15) [inline front-end/src/components/productView/ProductView.js]: Remove empty line 🙈

File: front-end/src/components/productView/ProductView.js [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+                }
+            }
+        }
+    }
+    
+`;
+
+const ProductView = (props) => {
+    const redirect = useHistory();
+    const productId = props.location.pathname.split('/')[1];
```
Comment (Svetloslav15) [inline front-end/src/components/productView/ProductView.js]: You can use useParams hook to get the id fron the url.
```suggestion
    const { id } = useParams();
```

File: front-end/src/services/productServices.js [hunk 18]
```diff
@@ -1,40 +1,75 @@
...
-        .catch(err => alert(err.message));
+            .then(res => {
+                alert(messages.success);
+            })
+            .catch(err => alert(err.message));
+    },
+    deleteProduct: (productId, redirect) => {
+        axios.delete(`${process.env.REACT_APP_BASE_URL}products/${productId}`)
+            .then(res => {
+                alert(productId + 'deleted')
```
Comment (Svetloslav15) [inline front-end/src/services/productServices.js]: Better message will be "ProductName was successfully deleted."

File: front-end/src/services/utils.js [hunk 19]
```diff
@@ -0,0 +1,5 @@
+export const utilFunc = {
+    stringifyUrl: (type, value) => {
```
Comment (Svetloslav15) [inline front-end/src/services/utils.js]: Could be simplified:
```suggestion
    stringifyUrl: (type, value) =>`&${type}=${value}`
```

File: front-end/src/components/gallery/Gallery.js (hunk 12 above)
Comment (Svetloslav15) [reply front-end/src/components/gallery/Gallery.js]: Small but, should be `price` with C

File: api/app.test.js [hunk 20]
```diff
@@ -2,10 +2,30 @@ const supertest = require('supertest');
...
 const app = require('./app');
 const request = supertest(app);
+const Product = require('./models/Product');
+
+const products = [{
+    title: 'Test',
+    category: 'Test',
+    size: 'Test',
+    color: 'Test',
+    price: 333,
```
Comment (Svetloslav15) [inline api/app.test.js:13]: You can add some more tests, for example adding a product with price negative value, when any of the props is missing, etc.

File: api/app.test.js [hunk 21]
```diff
@@ -16,10 +36,19 @@ async function removeAllCollections() {
...
 
-afterEach(async () => {
+afterAll(async () => {
     await removeAllCollections()
 })
 
+describe('Test Get products from database', () => {
+    test('should return up to 12 products', async () => {
+        const response = await request.get('/products')
+        expect(response.status).toBe(200)
```
Comment (Svetloslav15) [inline api/app.test.js]: You can use here the status codes from the status-codes.js

File: front-end/src/App.js [hunk 22]
```diff
@@ -12,6 +14,8 @@ function App() {
         <Switch>
           <Route exact path="/" component={Gallery}/>
           <Route exact path="/add" component={AddProduct}/>
+          <Route exact path="/:id" component={ProductView}/>
+          <Route exact path="/edit/:id" component={EditProduct} />
```
Comment (Svetloslav15) [inline front-end/src/App.js:29]: Should this line be before `<Route exact path="/:id" component={ProductView}/>`, because it will never be fired, I think?

Comment (Ivaylo15) [reply front-end/src/App.js:29]: It works but I can switch them.

File: api/controllers/product.js (hunk 7 above)
Comment (Ivaylo15) [reply api/controllers/product.js:18]: I use the id from the body and it is validated, never to be empty. If there are more validations needed I'll add them.

File: front-end/src/components/editProduct/EditProduct.js (hunk 10 above)
Comment (Ivaylo15) [reply front-end/src/components/editProduct/EditProduct.js:67]: I don't understand if I use only setTitle it returns an object.

File: api/controllers/product.js (hunk 7 above)
Comment (Svetloslav15) [reply api/controllers/product.js:18]: Okay cool! 😃

File: front-end/src/components/editProduct/EditProduct.js (hunk 10 above)
Comment (Svetloslav15) [reply front-end/src/components/editProduct/EditProduct.js:67]: It returns you the event, and then you have to get the value in the function with event.target.value.

File: front-end/src/App.js (hunk 22 above)
Comment (Svetloslav15) [reply front-end/src/App.js:29]: Weird, but okay if it works don't touch it. 🙈

File: front-end/src/components/editProduct/EditProduct.js (hunk 10 above)
Comment (Ivaylo15) [reply front-end/src/components/editProduct/EditProduct.js:67]: So I should write another function and pass it. Not directly use the arrow function in onChange.

Comment (Svetloslav15) [reply front-end/src/components/editProduct/EditProduct.js:67]: I will explain you tomorrow. 😄

File: api/constants/error-messages.js [hunk 23]
```diff
@@ -1,5 +1,5 @@
 module.exports = {
     MinLength: 'Minimum 3 characters required',
     EmptyField: 'Field can not be empty',
-    FieldNumber: 'Field must be a number',
+    FieldNumber: 'Field must be a positive number',
```
Comment (Svetloslav15) [inline api/constants/error-messages.js]: I think this should better be called, just to be more descriptive.
```suggestion
    InvalidPriceField: 'Field must be a positive number',
```

File: front-end/src/components/editProduct/EditProduct.js (hunk 10 above)
Comment (Ivaylo15) [reply front-end/src/components/editProduct/EditProduct.js:67]: Done something but not sure if it is right... I have extracted the value but it says that I'm changing from uncontrolled to controlled input. 100% I am not doing smth right.

**Review by Svetloslav15 — APPROVED** [review]
LGTM
